### Minimizing Fuel Costs - MILP

This is a basic mixed-integer linear problem for minimizing fuel costs. Example provided by [APMonitor](#https://www.youtube.com/watch?v=i8WS6HlE8qM)

Say we have two options for buying crude (expensive and cheap).

- The expensive crude is \$95 a barrel
- The cheap crude is \$80 a barrel
- The expensive crude produces 15 gallons of gasoline and 15 gallons of diesel
- The cheap crude produces 10 gallons of gasoline and 20 gallons of diesel
- We want to produce at least 100 gallons of gasoline and 160 gallons of diesel to meet current demand.
- We can only buy crude by the barrell.

In [1]:
#import our tools
import pyomo.environ as pyo

In [2]:
#initialize our concrete model
m = pyo.ConcreteModel()

#create our number of barrel variables (Expensive and cheap)
m.vEbbl = pyo.Var(within = pyo.NonNegativeIntegers)
m.vCbbl = pyo.Var(within = pyo.NonNegativeIntegers)

In [3]:
#create our cost minimization objective 
# cost = $95 * # of expensive bbls + $80 * # of cheap bbls
m.cost = pyo.Objective(expr = 95*m.vEbbl + 80*m.vCbbl, sense=pyo.minimize)

In [4]:
#Add our demand requirements (constraints if we want to meet current demand)
# gallons of gas  = 15 * # Exp bbl + 10 * # Cheap bbl (must be >= 100 gallons)
m.demandG = pyo.Constraint(expr = 15*m.vEbbl + 10*m.vCbbl >= 100)

# gallons of diesel  = 15 * # Exp bbl + 20 * # Cheap bbl (must be >= 160 gallons)
m.demandD = pyo.Constraint(expr = 15*m.vEbbl + 20*m.vCbbl >= 160)

In [5]:
#create solver and solve
results = pyo.SolverFactory('glpk').solve(m)

In [6]:
#view the results
results.write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 750.0
  Upper bound: 750.0
  Number of objectives: 1
  Number of constraints: 3
  Number of variables: 3
  Number of nonzeros: 5
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 3
      Number of created subproblems: 3
  Error rc: 0
  Time: 0.018649816513061523
# ----------------------------------------------------------
#   Solution Information
# ------------------------------------

Lowest cost is \$750, now let's se the number of barrels

In [7]:
m.pprint()

2 Var Declarations
    vCbbl : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :   7.0 :  None : False : False : NonNegativeIntegers
    vEbbl : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :   2.0 :  None : False : False : NonNegativeIntegers

1 Objective Declarations
    cost : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : 95*vEbbl + 80*vCbbl

2 Constraint Declarations
    demandD : Size=1, Index=None, Active=True
        Key  : Lower : Body                : Upper : Active
        None : 160.0 : 15*vEbbl + 20*vCbbl :  +Inf :   True
    demandG : Size=1, Index=None, Active=True
        Key  : Lower : Body                : Upper : Active
        None : 100.0 : 15*vEbbl + 10*vCbbl :  +Inf :   True

5 Declarations: vEbbl vCbbl cost demandG demandD


We see that we need 7 barrel of cheap crude and 2 barrels of expensive crude. If we plug those back into our demand solutions, we see that it produces us 170 gallons of Diesel and 100 gallons of gas